// dockerlint/lib/utils.js
// Generated by CoffeeScript 1.9.0
(function() {
  var args, sty, _base, _base1;

  args = require('subarg')(process.argv.slice(2), {
    alias: {
      d: 'debug',
      f: 'file',
      h: 'help',
      p: 'pedantic'
    }
  });

  sty = require('sty');

  if ((_base = String.prototype).beginsWith == null) {
    _base.beginsWith = function(s) {
      return this.slice(0, s.length) === s;
    };
  }

  if ((_base1 = String.prototype).endsWith == null) {
    _base1.endsWith = function(s) {
      return s === '' || this.slice(-s.length) === s;
    };
  }

  exports.notEmpty = function(s) {
    return !(s.trim() === '');
  };

  exports.log = function(level, msg) {
    switch (level) {
      case 'FATAL':
      case 5:
        console.error((sty.red('ERROR')) + ": " + msg);
        return process.exit(1);
      case 'ERROR':
      case 4:
        return console.error((sty.red('ERROR')) + ": " + msg);
      case 'WARN':
      case 3:
        console.warn((sty.red('WARN')) + ":  " + msg);
        if (args.pedantic) {
          return process.exit(1);
        }
        break;
      case 'INFO':
      case 2:
        return console.log((sty.green('INFO')) + ": " + msg);
      default:
        process.stdout.write((sty.blue('DEBUG')) + ":");
        return console.dir(msg);
    }
  };

}).call(this);
